### ISSUE TITLE
EfficientNet-b0 & MLP float point

### ISSUE THREAD

**Issue by qiminchen** [qiminchen/CoralNet#13]
I did some tests on both feature store and feature data type and found something very interesting.

1. Ways to store features
- JSON

    We currently store features of each image as JSON. Take 1 image with 200 points (from s294) as an example, the `.feature.json (vgg16 with a shape of 200x4096)`  I downloaded from the current `beta_export` is about `7.5MB`, the feature I extracted and saved as JSON `(efficientb0 with a shape of 200x1280)` is about `5.5MB`, both are `float64` if converting to the NumPy array.
    
    During the feature extraction, since the NumPy array is not JSON serializable, we cast the array `tolist()` and dump as JSON file for later classifier training. **(1)** No matter what data type `(float16, float32, float64)` the feature is, after storing as JSON, loading the JSON file, and converting to NumPy array, the data type is always `float64`. I am really confused about this and didn't find much information.

- NumPy `.npy`
    Instead I saved the features as `.npy` file and again for 1 image with 200 points (from s294), **(2)** `efficientb0 200x1280` where data type is `float16` is only `512.1kB`, `float32 - 1.0MB` and `float64 - 2.0MB`.

    But I am not very familiar with JSON so not sure the stability of numpy vs. JSON. Here is what I found on [StackOverflow](https://stackoverflow.com/questions/9619199/best-way-to-preserve-numpy-arrays-on-disk), you might be interested in the Ramdon array storage performance answer.

![storage](https://user-images.githubusercontent.com/45860234/96418487-ad526180-11a7-11eb-88a6-0c3886c05784.png)

2. Performance

Here is the summary table of `weighted accuracy, weighted f1-score and training time` across 26 test sources, refer to google sheet [EffNet & MLP float ablation](https://docs.google.com/spreadsheets/d/15NxI4AYrADqBB6KBQ-sL_ruEEf8mT10j_bq4w0tVhqI/edit?usp=sharing) for the performance of each test source.

**NOTE**: `float64` is the original data type where features are stored as JSON, while both `float32` and `float16` are directly stored as `.npy`.

|  Data type  |  Accuracy (%)  |  F1-Score  |  Training time (s)  |
|  :---:  |  :---:  |  :---:  |  :---:  |
| float64 | 77.58 | 0.7659 | 373.0312 |
| float32 | 77.75 | 0.7676 | 32.9096 |
| float16 | 77.63 | 0.7663 | 35.0504 |

I am very surprised that `float32` and `float16` would be so fast. I think this is mainly because `npy` retrieve is faster. As for the performance, I expected a slightly dropping in `float32` and `float16`, which did not happen, I think this might be because the MLP is too shallow to take advantage of the float precision. I might be wrong.

I will test the backend codebase on my local to verify if the training speed is really the case. If so, we might need to consider using a new data structure to store the features which will significantly save us a lot of resources. What do you guys think?

**Comment by qiminchen** [comment]
@kriegman @beijbom @StephenChan hey guys, i found something really interesting, can you take a look at it and let me know what you think?

**Comment by beijbom** [comment]
Thanks @qiminchen . Makes sense to me that train time would be faster but I'm surprised that the difference is so large. I guess `sklearn` has optimized the training code for 32 bit precision. This may also very across CPUs. I'm glad the accuracy wasn't affected. 

I was leaning towards serializing using protobuf but it doesn't look like proto3 supports float16, so we'd have to settle for float32. https://developers.google.com/protocol-buffers/docs/proto3.

We could also store directly using numpy, but we'd have to write something custom do handle the metadata as well as the row/col information. The full class definition is here: https://github.com/beijbom/pyspacer/blob/master/spacer/data_classes.py#L142. The idea is simply to overwrite the `.load()` and `.store()` methods of the superclass. @qiminchen : feel free to give this a go if you like.
